### DISCUSSION TITLE
Will it be possible to create an SSH tunnel to a Codespaces instance?

### DISCUSSION THREAD

**Discussion by paveloom** [community/community#25497] [Codespaces]
<p>For example, to connect via a local terminal, or to use SSHFS for easy access to project files.</p>

**Comment by lostintangent** [comment]
<p>Hey! This is something we’ve discussed adding support for, and it would be extremely helpful to learn more about your use case. Are you primarily  interested in this so that you could use another local desktop app against the files in the Codespace? (e.g. an image editor)</p>

**Comment by paveloom** [comment]
<p>Yes, for some file types it can be very convenient to use local applications (e.g. Adobe Acrobat DC for browsing/reading PDF files, WinRAR for viewing/unzipping archives, and so on). Quite often you need to quickly move files from the development environment to local storage, and vice versa (Visual Studio Code supports drag-n-drop in the latter case, but the former, as far as I know, is impossible without SSH). Finally, when I was writing this question, the main problem was the need to open HTML files in a browser, which is very easy to do if you use SSHFS.</p>
<p>Generally, the problems described can be partially or completely bypassed by using extensions inside the editor, however, these are crutches, and it would be better to have a more natural way of interaction.</p>

**Comment by lostintangent** [comment]
<p>Got it, yeah that makes total sense. Thanks for providing that extra detail! This is something we definitely want to enable, but we don’t have any concrete timelines to share yet.</p>
<p>As you said, the current workaround is to leverage an extension that augments the editor for the specific file type you want to manage. For viewing HTML files, you can checkout the <a href="https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer" rel="nofollow noopener">Live Server</a> extension and then forward the port that it starts up. Definitely not as smooth as it could be, but maybe enough to unblock you for now?</p>

**Comment by paveloom** [comment]
<p>Well, I have a slightly different approach to developing in the cloud right now, so it’s no longer a problem.</p>

**Comment by lostintangent** [comment]
<p>You mean you’re no longer using Codespaces? Or that you worked around this limitation in a different way?</p>

**Comment by paveloom** [comment]
<p>Yes, I don’t use it anymore. This problem is quite important for current projects, so for now, I keep Codespaces a bit on the sidelines.</p>

**Comment by Chuxel** [comment, marked as answer]
**UPDATE:** You can now set up a connection using the GitHub CLI.  See https://docs.github.com/en/codespaces/developing-in-codespaces/using-github-codespaces-with-github-cli#ssh-into-a-codespace and `gh codespace ssh --help` for setup options (like `--server-port` to specify a localhost port for use with other ssh clients).

-----

<p>See if this works for you:</p>
<ol>
<li>Create a codespace</li>
<li><a href="https://docs.github.com/en/github/developing-online-with-codespaces/connecting-to-your-codespace-from-visual-studio-code" rel="nofollow noopener">Connect to it using VS Code locally</a></li>
<li>Next, open a terminal (Crtl+Shift+`) in VS Code and run the following:<pre><code class="lang-auto">curl -sSL https://raw.githubusercontent.com/microsoft/vscode-dev-containers/master/script-library/sshd-debian.sh | sudo bash -s -- 2222 $(whoami) true random
</code></pre>
</li>
<li>Take note of the password that was generated and the SSH command.</li>
<li>Press F1 and select <strong>Codespaces: Forward Port</strong>
</li>
<li>Enter port 2222</li>
<li>Your codespace now has a running SSH server in it.  Use a local termal (or other tool) to connect to it using the command and password from step 3.  e.g.<pre><code class="lang-auto">ssh -p 2222 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null codespace@localhost
</code></pre>
</li>
</ol>
<p>The “-o” arguments are optional, but will prevent you from getting warnings or errors about known hosts when you do this from multiple codespaces.</p>
<p>Next time you connect, you can spin up the SSH server again by running: <code>/usr/local/share/sshd-init.sh</code> in a terminal and using the same command / password.</p>
<p>EDIT: Note that there is a right-click “Download” option in addition to dragging into the explorer to copy from local, but the script above will let you use SSHFS for local tool access.</p>

**Reply by shotgunner** [reply]
Thanks. but I get `Permission denied, please try again.` when I write password :(

**Reply by Chuxel** [reply]
As an update, you can now do this much more easily using the GitHub CLI - see https://docs.github.com/en/codespaces/developing-in-codespaces/using-github-codespaces-with-github-cli#ssh-into-a-codespace

This will forward the port locally automatically.  You can set the port using `--server-port` on the `gh codespaces ssh` command. The `--config` command can also be used to generate SSH config.

See `gh codespace ssh --help` for some examples!

**Reply by cross-hello** [reply]
Directly update account password:
`sudo passwd codespace`

**Comment by paveloom** [comment]
<p>Fantastic! Worked like a charm. Thank you very much, Chuck.</p>

**Comment by Chuxel** [comment]
<p>As a FYI - This script will soon be in the default image for Codespaces. Once that is there, you’ll can just run <code>sudo passwd $(whoami)</code> to set a password and then forward the SSH port. So it will soon be even easier.</p>
<p><strong>UPDATE:</strong> This is now live.  If you type <code>docker image ls</code> and see <code>0.17.0</code> or greater, a SSH server is there by default.</p>

**Comment by gtxg16** [comment]
How do I SSH into the codespace using SSH command and not gh cs ssh?

**Reply by Chuxel** [reply]
Use --server-port to specify the port the ssh server will be on localhost then connect.

**Reply by gtxg16** [reply]
`listen tcp 127.0.0.1:2222: bind: address already in use`
The SSH server is default on 2222

**Reply by gtxg16** [reply]
Moreover, which IP do I SSH to?

**Reply by Chuxel** [reply]
You can also add something to your SSH config file to make this easier using `gh codespace ssh --config` e.g. `gh codespace ssh --config >> ~/.ssh/config`

The command is basically, `ssh -p 2222 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null codespace@localhost` as described above Update `codespace` to the user in the Codespace.

**Comment by atepavicharov** [comment]
I don't get it, why localhost here `ssh -p 2222 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null codespace@localhost`. If I want to ssh into a codespace without using the gh cli how can I do that ?

**Reply by shotgunner** [reply]
probably because they don't want to use it as VPN ssh tunneling (with `-D ssh option`)

**Comment by vidjuheffex** [comment]
anyone care to summarize what to do to accomplish this?
